app/Helpers/database/user-activity.php

```diff
@@ -177,87 +161,46 @@ function postActivity(string $userIn, int $type, ?int $data = null, ?int $data2
                  */
             }
 
-            $query .= "(NOW(), $type, '$user', '$gameID', NULL)";
+            $activity->data = (string) $gameID;
             break;
 
         case ActivityType::UploadAchievement:
         case ActivityType::EditAchievement:
         case ActivityType::OpenedTicket:
         case ActivityType::ClosedTicket:
-            $query .= "(NOW(), $type, '$user', '$data', NULL)";
+            $activity->data = (string) $data;
             break;
 
         case ActivityType::CompleteGame:
         case ActivityType::NewLeaderboardEntry:
         case ActivityType::ImprovedLeaderboardEntry:
-            $query .= "(NOW(), $type, '$user', '$data', '$data2')";
+            $activity->data = (string) $data;
+            $activity->data2 = (string) $data2;
             break;
     }
 
-    $dbResult = mysqli_query($db, $query);
-    if (!$dbResult) {
-        log_sql_fail();
-
-        return false;
-    }
+    $activity->save();
 
     if ($type == ActivityType::StartedPlaying) {
-        // have to do this after the query is executed to prevent a race condition where
-        // it may get re-cached before the query finishes
-        expireRecentlyPlayedGames($user);
+        // have to do this after the activity is saved to prevent a race condition where
+        // it may get re-cached before the activity is committed.
+        expireRecentlyPlayedGames($user->User);
     }
 
-    /**
-     * Update UserAccount
-     */
-    $newActID = mysqli_insert_id($db);
-    $query = "UPDATE UserAccounts AS ua SET ua.LastActivityID = $newActID, ua.LastLogin = NOW() WHERE ua.User = '$user'";
-    $dbResult = s_mysql_query($query);
-
-    if (!$dbResult) {
-        log_sql_fail();
-
-        return false;
-    }
+    // update UserAccount
+    $user->LastLogin = Carbon::now();
+    $user->LastActivityID = $activity->ID;
+    $user->timestamps = false;
+    $user->save();
 
     return true;
 }
 
-function userActivityPing(string $user): bool
-{
-    $username = validateUsername($user);
-    if (!$username) {
-        return false;
-    }
-
-    return legacyDbStatement(
-        'UPDATE UserAccounts AS ua SET ua.LastLogin = NOW() WHERE ua.User = :username',
-        ['username' => $username]
-    );
-}
-
-function UpdateUserRichPresence(string $user, int $gameID, string $presenceMsg): bool
+function UpdateUserRichPresence(User &$user, int $gameID, string $presenceMsg): void
```

Comment:
I'd go with returning the user model after saving the changes instead of pass-by-reference.
Saving here would be the expected side effect going by `update` in the function name.

Reply:
This isn't actually saving. The save occurs in the two places that call it as they make additional changes to the User object.

PHP documentation indicates that all objects are already passed by reference, so the & is redundant. I'll remove it.